x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots;
+
+import org.elasticsearch.cluster.routing.RecoverySource;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.cluster.routing.allocation.AllocateUnassignedDecision;
+import org.elasticsearch.cluster.routing.allocation.AllocationDecision;
+import org.elasticsearch.cluster.routing.allocation.ExistingShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.FailedShard;
+import org.elasticsearch.cluster.routing.allocation.NodeAllocationResult;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SearchableSnapshotAllocator implements ExistingShardsAllocator {
+
+    static final String ALLOCATOR_NAME = "searchable_snapshot_allocator";
+
+    @Override
+    public void beforeAllocation(RoutingAllocation allocation) {
+    }
+
+    @Override
+    public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {
+    }
+
+    @Override
+    public void allocateUnassigned(RoutingAllocation allocation, ShardRouting shardRouting,
+                                   RoutingNodes.UnassignedShards.UnassignedIterator iterator) {
+        final AllocateUnassignedDecision allocateUnassignedDecision = decideAllocation(allocation, shardRouting);
+        assert allocateUnassignedDecision.isDecisionTaken();
+
+        if (allocateUnassignedDecision.getAllocationDecision() == AllocationDecision.YES) {
+            if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.EXISTING_STORE) {
+                // we don't care what the allocation ID is since we know that these shards cannot really be stale, so we can
+                // safely ignore the allocation ID with a forced-stale allocation
+                iterator.updateUnassigned(shardRouting.unassignedInfo(),
+                    RecoverySource.ExistingStoreRecoverySource.FORCE_STALE_PRIMARY_INSTANCE, allocation.changes());
+            }
+            iterator.initialize(allocateUnassignedDecision.getTargetNode().getId(), null, 0L, allocation.changes());
+        } else {
+            iterator.removeAndIgnore(allocateUnassignedDecision.getAllocationStatus(), allocation.changes());
+        }
+    }
+
+    private static AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation, ShardRouting shardRouting) {
+        assert shardRouting.unassigned();
```

Comment:
I think the decision made here effectively bypasses the balancer. I assume you want to tackle that in a follow-up? Would be nice to know if the intention is to invoke balancer somehow here or if the code here will be changed to only consider the nodes that already have data that can be reused?

Reply:
Yes, the actual allocator's behaviour is not really the focus of this PR so I've just done something very simple. To be discussed in the follow-up, but indeed I agree that we will allow shards to fall through to the balancer if we can't find them a good home here.